File: android/app/src/test/java/com/quartier/quartier/SupabaseTests.kt
package com.quartier.quartier

import com.quartier.quartier.auth.AuthManager
import com.quartier.quartier.auth.AuthManagerImpl
import com.quartier.quartier.auth.LoginViewModel
import com.quartier.quartier.database.AuthRepository
import com.quartier.quartier.database.AuthRepositoryImpl
import com.quartier.quartier.database.Connection
import com.quartier.quartier.database.ConnectionRequestResult
import com.quartier.quartier.database.ConnectionsDatabase
import com.quartier.quartier.database.ConnectionsRepository
import com.quartier.quartier.database.Socials
import com.quartier.quartier.database.SocialsDatabase
import com.quartier.quartier.database.SocialsRepository
import com.quartier.quartier.database.UserDatabase
import com.quartier.quartier.database.UserRepository
import com.quartier.quartier.mock_models.MockAuthManager
import io.github.jan.supabase.auth.auth
import io.ktor.http.Url
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.StandardTestDispatcher
import kotlinx.coroutines.test.resetMain
import kotlinx.coroutines.test.runTest
import kotlinx.coroutines.test.setMain
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import java.net.URL
import kotlin.math.log
import kotlin.test.assertContains
import kotlin.test.assertEquals
import kotlin.uuid.ExperimentalUuidApi
import kotlin.uuid.Uuid

@OptIn(ExperimentalCoroutinesApi::class)
@RunWith(RobolectricTestRunner::class)
class SupabaseTests {
    private lateinit var authManager: AuthManager
    private lateinit var authRepo: AuthRepository
    private lateinit var userRepo: UserRepository
    private lateinit var connectRepo: ConnectionsRepository
    private lateinit var socialsRepo: SocialsRepository

    private val testDispatcher = StandardTestDispatcher()

    @Before
    fun setUp() {
        Dispatchers.setMain(testDispatcher)
        authManager = AuthManagerImpl()
        authRepo = AuthRepositoryImpl()
        userRepo = UserDatabase(authRepo)
        connectRepo = ConnectionsDatabase(authRepo)
        socialsRepo = SocialsDatabase(authRepo)
    }

    @After
    fun tearDown() {
        Dispatchers.resetMain()
    }

    @OptIn(ExperimentalUuidApi::class)
    suspend fun createUser() : String { //Helper function to create a user for tests
        val email = Uuid.random().toString()+"@example.com"
        authManager.emailSignUp("Test User", email, "Password123")
        authRepo.updateUserId(supabase.auth.currentUserOrNull()!!.id)
        return email
    }

    suspend fun loginAs(email: String) { //Helper function to login for tests
        authManager.emailSignIn(email, "Password123")
        authRepo.updateUserId(supabase.auth.currentUserOrNull()!!.id)
    }

    // -------- Auth Tests --------
    @Test fun signup_valid() = runTest { //Create a user and ensure it is signed up
        val email = createUser()
        testScheduler.advanceUntilIdle()

        assert(supabase.auth.currentUserOrNull()?.id != null)
    }

    @Test fun signIn_valid() = runTest { //Create a user, log out, then make sure login works.
        val email = createUser()
        supabase.auth.signOut()
        authManager.emailSignIn(email, "Password123")
        testScheduler.advanceUntilIdle()

        assert(supabase.auth.currentUserOrNull()?.id != null)
    }

    // -------- User Table Tests --------
    @Test fun getUser_profile() = runTest { //Get a user's profile
        createUser()
        val user = userRepo.getUser()

        assertEquals("Test User", user.name) //Default test user name
    }

    @Test fun getUsers() = runTest { //Create 2 users and get their profiles by id
        createUser()
        val user1 = userRepo.getUser()

        createUser()
        val user2 = userRepo.getUser()

        val users = userRepo.getUsers(listOf(user1.id, user2.id))
        testScheduler.advanceUntilIdle()

        assertContains(users, user1)
        assertContains(users, user2)
    }

    @Test fun updateUserProfile() = runTest { //Make sure a user's profile is edited properly
        createUser()

        userRepo.updateUser("New Name", "Amazing Job Title")
        val updatedUser = userRepo.getUser()
        testScheduler.advanceUntilIdle()

        assertEquals("New Name", updatedUser.name)
        assertEquals("Amazing Job Title", updatedUser.job)
    }

    @OptIn(ExperimentalUuidApi::class)
    @Test fun uploadUserImage() = runTest { //Make sure a pfp can be updated properly
        createUser()

        val imageUrl = "https://picsum.photos/400/400.jpg"
        val url = URL(imageUrl)
        val data = url.readBytes() //Read a random image and upload it

        userRepo.uploadPfp(Uuid.random().toString()+".jpg", data)
        val profile = userRepo.getUser()

        testScheduler.advanceUntilIdle()

        assert(profile.pfp_url != null)
    }

    // -------- Connections Table Tests --------
    @Test fun requestConnection_andGet() = runTest { //Create connections and get them all
        createUser()
        val id1 = userRepo.getUser().id

        createUser()
        val id2 = userRepo.getUser().id
        connectRepo.requestConnection(id1)

        val connection = connectRepo.getConnectionWithUser(id1)
        testScheduler.advanceUntilIdle()

        assertEquals(id2, connection?.requested_by)
        assertEquals(id1, connection?.requested_for)
        assertEquals("pending", connection?.status)
    }

    @Test fun tryConnect_withSelf() = runTest { //User attempts to connect with themselves
        createUser()
        val id = userRepo.getUser().id

        val result = connectRepo.requestConnection(id)
        testScheduler.advanceUntilIdle()

        assertEquals(ConnectionRequestResult.CannotConnectWithSelf, result)
    }

    @Test fun acceptConnection_andGet() = runTest { //Send a request, accept it, and get the connection
        val email = createUser()
        val id1 = userRepo.getUser().id

        createUser()
        val id2 = userRepo.getUser().id
        connectRepo.requestConnection(id1)

        loginAs(email)
        connectRepo.acceptConnection(id2)

        val connection = connectRepo.getConnectionWithUser(id2)
        testScheduler.advanceUntilIdle()

        assertEquals(id2, connection?.requested_by)
        assertEquals(id1, connection?.requested_for)
        assertEquals("accepted", connection?.status)
    }

    @Test fun deleteConnection() = runTest { //Make sure a connection is deleted properly
        val email = createUser()
        val id1 = userRepo.getUser().id

        createUser()
        val id2 = userRepo.getUser().id
        connectRepo.requestConnection(id1)

        loginAs(email)
        connectRepo.deleteConnection(id2)

        val connection = connectRepo.getConnectionWithUser(id2)
        testScheduler.advanceUntilIdle()

        assertEquals(null, connection)
    }

    @Test fun getAllConnections() = runTest { //Get all of a user's connections
        val email = createUser()
        val id1 = userRepo.getUser().id

        createUser()
        val id2 = userRepo.getUser().id
        connectRepo.requestConnection(id1)

        createUser()
        val id3 = userRepo.getUser().id
        connectRepo.requestConnection(id1)

        loginAs(email)
        val connections = connectRepo.getConnections()
        testScheduler.advanceUntilIdle()

        assertContains(connections, Connection(id2, id1, "pending"))
        assertContains(connections, Connection(id3, id1, "pending"))
    }

    // -------- Socials Table Tests --------
    @Test fun upsert_getSocials() = runTest { //Edit the socials and get the changes
        createUser()
        socialsRepo.upsertSocials("https://www.linkedin.com/in/user")

        val socials = socialsRepo.getUserSocials()
        assertEquals("https://www.linkedin.com/in/user", socials.linkedin_url)
    }

    @Test fun getConnectedSocials() = runTest { //Get all connections' socials
        val email = createUser()
        val id1 = userRepo.getUser().id

        createUser()
        val id2 = userRepo.getUser().id
        socialsRepo.upsertSocials("https://www.linkedin.com/in/user2")
        connectRepo.requestConnection(id1)

        createUser()
        val id3 = userRepo.getUser().id
        socialsRepo.upsertSocials("https://www.linkedin.com/in/user3")
        connectRepo.requestConnection(id1)

        loginAs(email)
        connectRepo.acceptConnection(id2)
        connectRepo.acceptConnection(id3)

        val socials = socialsRepo.getUserSocialsList()
        testScheduler.advanceUntilIdle()

        assertContains(socials, Socials(id2, "https://www.linkedin.com/in/user2"))
        assertContains(socials, Socials(id3, "https://www.linkedin.com/in/user3"))
    }
}

File: ios/QuartierTests/SupabaseTests.swift
//
//  SupabaseTests.swift
//  Quartier
//
//  Created by Emile Turcotte on 2025-08-08.
//

import Foundation
import Testing
@testable import Quartier

func createUser() async throws -> String { //Create a user for tests and return the uuid
    let auth = AuthManagerImpl.shared
    
    let email = UUID().uuidString+"@example.com"
    
    try await auth.signup(email: email, pwd: "Password123", name: "Test User")
    return email
}

@Suite(.serialized) struct SupabaseTests { //All the supabase tests are in the same class since they need to be run one by one to work
    /*--------------------------------
               AUTH TESTS
    --------------------------------*/
    
    @Test func signup_valid() async throws { //Try creating a new user
        let auth = AuthManagerImpl.shared
        
        await #expect(throws: Never.self) {
            try await auth.signup(email: UUID().uuidString+"@example.com", pwd: "Password123", name: "Test User")
        }
    }
    
    @Test func signin_valid() async throws { //Create a new user and then login
        let auth = AuthManagerImpl.shared
        
        let email = try await createUser()
        
        await #expect(throws: Never.self) {
            try await auth.login(email: email, pwd: "Password123")
        }
    }
    
    /*--------------------------------
            USER TABLE TESTS
    --------------------------------*/
    
    @Test func getUser_profile() async throws { //Try getting a user's profile
        let db = UserDatabase.shared
        
        let _ = try await createUser() //Create a random user and get their profile
        let user = try await db.getUser()
        #expect(user!.name == "Test User") //By default, the name should be "Test User" (see AuthTests for the createUser() function)
    }
    
    @Test func getUsers() async throws { //Create 2 users, and get their profiles by user id
        let db = UserDatabase.shared
        
        let _ = try await createUser() //Create user 1
        let user1 = try await db.getUser()
        
        let _ = try await createUser() //Create user 2
        let user2 = try await db.getUser()
        
        let users = try await db.getUsers(ids: [user1!.id, user2!.id])
        #expect(users.contains([user1!, user2!]))
    }
    
    @Test func updateUserProfile() async throws { //Edit a user's profile
        let db = UserDatabase.shared
        
        let _ = try await createUser()
    
        try await db.updateUser(name: "New Name", jobTitle: "Amazing Job Title") //Change their profile
        let newUser = try await db.getUser() //Get the profile to ensure the changes have been applied
        
        #expect(newUser!.name == "New Name")
        #expect(newUser!.job == "Amazing Job Title")
    }
    
    @Test func uploadUserImage() async throws {
        let db = UserDatabase.shared
        
        let _ = try await createUser()
        
        let imageUrl = "https://picsum.photos/400/400.jpg" //Placeholder image
        let imageData = try Data(contentsOf: URL(string: imageUrl)!)
        
        try await db.uploadPfp(fileName: UUID().uuidString+".jpg", imageData: imageData) //Upload image
        let profile = try await db.getUser()
        
        #expect(profile!.pfp_url != nil) //Make sure there is a link pointing to the image
    }
    
    /*--------------------------------
         CONNECTIONS TABLE TESTS
    --------------------------------*/
    @Test func requestConnection_andGet() async throws { //Get all of a user's connections
        let userDB = UserDatabase.shared
        let db = ConnectionsDatabase.shared
        
        let _ = try await createUser() //Create a random user and get their id
        let user1ID = try await userDB.getUser()!.id
        
        let _ = try await createUser() //Create another user and get their id
        let user2ID = try await userDB.getUser()!.id
        let _ = try await db.requestConnection(requestedID: user1ID) //Make the connection
        
        let connection = try await db.getConnectionWithUser(requestedID: user1ID) //Check if the connection was made properly
        #expect(connection?.requested_by == user2ID)
        #expect(connection?.requested_for == user1ID)
        #expect(connection?.status == "pending")
    }
    
    @Test func requestConnection_withSelf() async throws { //Get all of a user's connections
        let userDB = UserDatabase.shared
        let db = ConnectionsDatabase.shared
        
        let _ = try await createUser() //Create a random user and get their id
        let user1ID = try await userDB.getUser()!.id
        
        let result = try await db.requestConnection(requestedID: user1ID) //Make the connection
        
        #expect(result == .cannotConnectWithSelf)
    }
    
    @Test func acceptConnection_andGet() async throws {
        let auth = AuthManagerImpl.shared
        let userDB = UserDatabase.shared
        let db = ConnectionsDatabase.shared
        
        let email = try await createUser() //Create a random user and get their email
        let user1ID = try await userDB.getUser()!.id
        
        let _ = try await createUser()
        let user2ID = try await userDB.getUser()!.id //Create another user and get their id
        let _ = try await db.requestConnection(requestedID: user1ID) //Make the connection
        
        try await auth.login(email: email, pwd: "Password123")
        let _ = try await db.acceptConnection(requestedID: user2ID)
        
        let connection = try await db.getConnectionWithUser(requestedID: user2ID) //Check if the connection was made properly
        #expect(connection?.requested_by == user2ID)
        #expect(connection?.requested_for == user1ID)
        #expect(connection?.status == "accepted")
    }
    
    @Test func deleteConnection_andGet() async throws {
        let auth = AuthManagerImpl.shared
        let userDB = UserDatabase.shared
        let db = ConnectionsDatabase.shared
        
        let email = try await createUser() //Create a random user and get their email
        let user1ID = try await userDB.getUser()!.id
        
        let _ = try await createUser()
        let user2ID = try await userDB.getUser()!.id //Create another user and get their id
        let _ = try await db.requestConnection(requestedID: user1ID) //Make the connection
        
        try await auth.login(email: email, pwd: "Password123")
        let _ = try await db.deleteConnection(requestedID: user2ID)
        
        let connection = try await db.getConnectionWithUser(requestedID: user2ID) //Check if the connection was made properly
        #expect(connection == nil)
    }
    
    @Test func getConnections() async throws {
        let auth = AuthManagerImpl.shared
        let userDB = UserDatabase.shared
        let db = ConnectionsDatabase.shared
        
        let email = try await createUser() //Create a random user and get their email
        let user1ID = try await userDB.getUser()!.id
        
        let _ = try await createUser()
        let user2ID = try await userDB.getUser()!.id //Create another user and get their id
        let _ = try await db.requestConnection(requestedID: user1ID) //Make the connection
        
        let _ = try await createUser()
        let user3ID = try await userDB.getUser()!.id //Create another user and get their id
        let _ = try await db.requestConnection(requestedID: user1ID) //Make the connection
        
        try await auth.login(email: email, pwd: "Password123")
        let connections = try await db.getConnections()
        #expect(connections.contains(Connection(requested_by: user2ID, requested_for: user1ID, status: "pending")))
        #expect(connections.contains(Connection(requested_by: user3ID, requested_for: user1ID, status: "pending")))
    }
    
    /*--------------------------------
           SOCIALS TABLE TESTS
    --------------------------------*/
    @Test func upsert_and_getSocials() async throws {
        let db = SocialsDatabase.shared
        
        let _ = try await createUser()
        try await db.upsertSocials(linkedInUrl: "https://www.linkedin.com/in/user")
        
        let socials = try await db.getUserSocials()
        #expect(socials!.linkedin_url == "https://www.linkedin.com/in/user")
    }
    
    @Test func getConnectedSocials() async throws {
        let auth = AuthManagerImpl.shared
        let userDB = UserDatabase.shared
        let connDB = ConnectionsDatabase.shared
        let socialsDB = SocialsDatabase.shared
        
        let email = try await createUser() //Create a random user and get their email
        let user1ID = try await userDB.getUser()!.id
        
        let _ = try await createUser()
        let user2ID = try await userDB.getUser()!.id //Create another user and get their id
        try await socialsDB.upsertSocials(linkedInUrl: "https://www.linkedin.com/in/user2")
        let _ = try await connDB.requestConnection(requestedID: user1ID) //Make the connection
        
        let _ = try await createUser()
        let user3ID = try await userDB.getUser()!.id //Create another user and get their id
        try await socialsDB.upsertSocials(linkedInUrl: "https://www.linkedin.com/in/user3")
        let _ = try await connDB.requestConnection(requestedID: user1ID) //Make the connection
        
        try await auth.login(email: email, pwd: "Password123")
        let _ = try await connDB.acceptConnection(requestedID: user2ID)
        let _ = try await connDB.acceptConnection(requestedID: user3ID)
        
        let socials = try await socialsDB.getConnectedSocials()
        #expect(socials.contains([
            Socials(id: user2ID, linkedin_url: "https://www.linkedin.com/in/user2"),
            Socials(id: user3ID, linkedin_url: "https://www.linkedin.com/in/user3"),
        ]))
    }
}
